remotes/docker/config/hosts.go

```diff
@@ -283,202 +284,203 @@ type hostFileConfig struct {
 	Capabilities []string `toml:"capabilities"`
 
 	// CACert can be a string or an array of strings
-	CACert toml.Primitive `toml:"ca"`
+	CACert interface{} `toml:"ca"`
 
 	// TODO: Make this an array (two key types, one for pairs (multiple files), one for single file?)
-	Client toml.Primitive `toml:"client"`
+	Client interface{} `toml:"client"`
 
 	SkipVerify *bool `toml:"skip_verify"`
 
-	Header map[string]toml.Primitive `toml:"header"`
+	Header map[string]interface{} `toml:"header"`
 
 	// API (default: "docker")
 	// API Version (default: "v2")
 	// Credentials: helper? name? username? alternate domain? token?
 }
 
-type configFile struct {
-	// hostConfig holds defaults for all hosts as well as
-	// for the default server
-	hostFileConfig
+func parseHostsFile(baseDir string, b []byte) ([]hostConfig, error) {
+	tree, err := toml.LoadBytes(b)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse TOML")
+	}
+
+	// HACK: we want to keep toml parsing structures private in this package, however go-toml ignores private embedded types.
+	// so we remap it to a public type within the func body, so technically it's public, but not possible to import elsewhere.
+	type HostFileConfig = hostFileConfig
 
-	// Server specifies the default server. When `host` is
-	// also specified, those hosts are tried first.
-	Server string `toml:"server"`
+	c := struct {
+		HostFileConfig
+		// Server specifies the default server. When `host` is
+		// also specified, those hosts are tried first.
+		Server string `toml:"server"`
+		// HostConfigs store the per-host configuration
+		HostConfigs map[string]hostFileConfig `toml:"host"`
+	}{}
 
-	// HostConfigs store the per-host configuration
-	HostConfigs map[string]hostFileConfig `toml:"host"`
-}
+	var (
+		hosts []hostConfig
+	)
 
-func parseHostsFile(ctx context.Context, baseDir string, b []byte) ([]hostConfig, error) {
-	var c configFile
-	md, err := toml.Decode(string(b), &c)
+	if err := tree.Unmarshal(&c); err != nil {
+		return nil, err
+	}
+
+	// Parse root host config
+	parsed, err := parseHostConfig(c.Server, baseDir, c.HostFileConfig)
 	if err != nil {
 		return nil, err
 	}
+	hosts = append(hosts, parsed)
 
-	var orderedHosts []string
-	for _, key := range md.Keys() {
-		if len(key) >= 2 {
-			if key[0] == "host" && (len(orderedHosts) == 0 || orderedHosts[len(orderedHosts)-1] != key[1]) {
-				orderedHosts = append(orderedHosts, key[1])
-			}
+	// Parse hosts array
+	for host, config := range c.HostConfigs {
+		parsed, err := parseHostConfig(host, baseDir, config)
+		if err != nil {
+			return nil, err
 		}
+		hosts = append(hosts, parsed)
```

Comment:
The ordering here is important and I don't see where it is preserved. Is that part of the test coverage?